File: concept/type/ThingTypeImpl.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.concept.type;

import com.vaticle.typedb.client.api.TypeDBTransaction;
import com.vaticle.typedb.client.api.concept.type.AttributeType;
import com.vaticle.typedb.client.api.concept.type.RoleType;
import com.vaticle.typedb.client.api.concept.type.ThingType;
import com.vaticle.typedb.client.common.Label;
import com.vaticle.typedb.client.concept.ConceptManagerImpl;
import com.vaticle.typedb.client.concept.thing.ThingImpl;
import com.vaticle.typedb.client.jni.Transitivity;
import com.vaticle.typedb.client.jni.ValueType;
import com.vaticle.typeql.lang.common.TypeQLToken;

import java.util.Set;
import java.util.stream.Stream;

import static com.vaticle.typedb.client.jni.typedb_client.concept_is_attribute_type;
import static com.vaticle.typedb.client.jni.typedb_client.concept_is_entity_type;
import static com.vaticle.typedb.client.jni.typedb_client.concept_is_relation_type;
import static com.vaticle.typedb.client.jni.typedb_client.concept_is_root_thing_type;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_delete;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_get_label;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_get_owns;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_get_owns_overridden;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_get_plays;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_get_plays_overridden;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_get_syntax;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_is_abstract;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_is_deleted;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_set_abstract;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_set_label;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_set_owns;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_set_plays;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_unset_abstract;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_unset_owns;
import static com.vaticle.typedb.client.jni.typedb_client.thing_type_unset_plays;
import static com.vaticle.typedb.common.collection.Collections.set;
import static java.util.Collections.emptySet;

public abstract class ThingTypeImpl extends TypeImpl implements ThingType {

    ThingTypeImpl(com.vaticle.typedb.client.jni.Concept concept) {
        super(concept);
    }

    public static ThingTypeImpl of(com.vaticle.typedb.client.jni.Concept concept) {
        if (concept_is_entity_type(concept)) return new EntityTypeImpl(concept);
        else if (concept_is_relation_type(concept)) return new RelationTypeImpl(concept);
        else if (concept_is_attribute_type(concept)) return new AttributeTypeImpl(concept);
        else if (concept_is_root_thing_type(concept)) return new Root(concept);
        return null; // FIXME throw
    }

    @Override
    public final boolean isRoot() {
        return false; // FIXME
//        return thing_type_is_root(concept);
    }

    @Override
    public final boolean isAbstract() {
        return thing_type_is_abstract(concept);
    }

    @Override
    public Label getLabel() {
        return Label.of(thing_type_get_label(concept));
    }

    @Override
    public void delete(TypeDBTransaction transaction) {
        thing_type_delete(((ConceptManagerImpl) transaction.concepts()).transaction, concept);
    }

    @Override
    public boolean isDeleted(TypeDBTransaction transaction) {
        return thing_type_is_deleted(((ConceptManagerImpl) transaction.concepts()).transaction, concept);
    }

    @Override
    public final void setLabel(TypeDBTransaction transaction, String newLabel) {
        thing_type_set_label(((ConceptManagerImpl) transaction.concepts()).transaction, concept, newLabel);
    }

    @Override
    public abstract ThingTypeImpl getSupertype(TypeDBTransaction transaction);

    @Override
    public abstract Stream<? extends ThingTypeImpl> getSupertypes(TypeDBTransaction transaction);

    @Override
    public abstract Stream<? extends ThingTypeImpl> getSubtypes(TypeDBTransaction transaction);

    @Override
    public abstract Stream<? extends ThingTypeImpl> getSubtypesExplicit(TypeDBTransaction transaction);

    @Override
    public abstract Stream<? extends ThingImpl> getInstances(TypeDBTransaction transaction);

    @Override
    public abstract Stream<? extends ThingImpl> getInstancesExplicit(TypeDBTransaction transaction);

    @Override
    public final void setAbstract(TypeDBTransaction transaction) {
        thing_type_set_abstract(((ConceptManagerImpl) transaction.concepts()).transaction, concept);
    }

    @Override
    public final void unsetAbstract(TypeDBTransaction transaction) {
        thing_type_unset_abstract(((ConceptManagerImpl) transaction.concepts()).transaction, concept);
    }

    @Override
    public final void setPlays(TypeDBTransaction transaction, RoleType roleType) {
        thing_type_set_plays(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((RoleTypeImpl) roleType).concept, null);
    }

    @Override
    public final void setPlays(TypeDBTransaction transaction, RoleType roleType, RoleType overriddenRoleType) {
        thing_type_set_plays(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((RoleTypeImpl) roleType).concept, ((RoleTypeImpl) overriddenRoleType).concept);
    }

    @Override
    public void setOwns(TypeDBTransaction transaction, AttributeType attributeType) {
        setOwns(transaction, attributeType, null, emptySet());
    }

    @Override
    public void setOwns(TypeDBTransaction transaction, AttributeType attributeType, Set<TypeQLToken.Annotation> annotations) {
        setOwns(transaction, attributeType, null, annotations);
    }

    @Override
    public void setOwns(TypeDBTransaction transaction, AttributeType attributeType, AttributeType overriddenType) {
        setOwns(transaction, attributeType, overriddenType, emptySet());
    }

    @Override
    public final void setOwns(TypeDBTransaction transaction, AttributeType attributeType, AttributeType overriddenType, Set<TypeQLToken.Annotation> annotations) {
        thing_type_set_owns(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((AttributeTypeImpl) attributeType).concept, overriddenType != null ? ((AttributeTypeImpl) overriddenType).concept : null, annotations.stream().map(TypeQLToken.Annotation::toString).toArray(String[]::new));
    }

    @Override
    public final Stream<RoleTypeImpl> getPlays(TypeDBTransaction transaction) {
        return getPlays(transaction, Transitivity.Transitive);
    }

    @Override
    public final Stream<RoleTypeImpl> getPlaysExplicit(TypeDBTransaction transaction) {
        return getPlays(transaction, Transitivity.Explicit);
    }

    private Stream<RoleTypeImpl> getPlays(TypeDBTransaction transaction, Transitivity transitivity) {
        return thing_type_get_plays(((ConceptManagerImpl) transaction.concepts()).transaction, concept, transitivity).stream().map(RoleTypeImpl::new);
    }

    @Override
    public RoleTypeImpl getPlaysOverridden(TypeDBTransaction transaction, RoleType roleType) {
        com.vaticle.typedb.client.jni.Concept res = thing_type_get_plays_overridden(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((RoleTypeImpl) roleType).concept);
        if (res != null) return new RoleTypeImpl(res);
        else return null;
    }

    @Override
    public Stream<AttributeTypeImpl> getOwns(TypeDBTransaction transaction) {
        return getOwns(transaction, Transitivity.Transitive, set());
    }

    @Override
    public Stream<AttributeTypeImpl> getOwns(TypeDBTransaction transaction, ValueType valueType) {
        return getOwns(transaction, valueType, Transitivity.Transitive, set());
    }

    @Override
    public Stream<AttributeTypeImpl> getOwns(TypeDBTransaction transaction, Set<TypeQLToken.Annotation> annotations) {
        return getOwns(transaction, Transitivity.Transitive, annotations);
    }

    @Override
    public final Stream<AttributeTypeImpl> getOwns(TypeDBTransaction transaction, ValueType valueType, Set<TypeQLToken.Annotation> annotations) {
        return getOwns(transaction, valueType, Transitivity.Transitive, annotations);
    }

    @Override
    public Stream<? extends AttributeType> getOwnsExplicit(TypeDBTransaction transaction) {
        return getOwns(transaction, Transitivity.Explicit, set());
    }

    @Override
    public Stream<? extends AttributeType> getOwnsExplicit(TypeDBTransaction transaction, ValueType valueType) {
        return getOwns(transaction, valueType, Transitivity.Explicit, set());
    }

    @Override
    public Stream<? extends AttributeType> getOwnsExplicit(TypeDBTransaction transaction, Set<TypeQLToken.Annotation> annotations) {
        return getOwns(transaction, Transitivity.Explicit, annotations);
    }

    @Override
    public Stream<? extends AttributeType> getOwnsExplicit(TypeDBTransaction transaction, ValueType valueType, Set<TypeQLToken.Annotation> annotations) {
        return getOwns(transaction, valueType, Transitivity.Explicit, annotations);
    }

    private Stream<AttributeTypeImpl> getOwns(TypeDBTransaction transaction, Transitivity transitivity, Set<TypeQLToken.Annotation> annotations) {
        return getOwns(transaction, null, transitivity, annotations);
    }

    private Stream<AttributeTypeImpl> getOwns(TypeDBTransaction transaction, ValueType valueType, Transitivity transitivity, Set<TypeQLToken.Annotation> annotations) {
        return thing_type_get_owns(((ConceptManagerImpl) transaction.concepts()).transaction, concept, valueType, transitivity, annotations.stream().map(TypeQLToken.Annotation::toString).toArray(String[]::new)).stream().map(AttributeTypeImpl::new);
    }

    @Override
    public AttributeTypeImpl getOwnsOverridden(TypeDBTransaction transaction, AttributeType attributeType) {
        com.vaticle.typedb.client.jni.Concept res = thing_type_get_owns_overridden(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((AttributeTypeImpl) attributeType).concept);
        if (res != null) return new AttributeTypeImpl(res);
        else return null;
    }

    @Override
    public final void unsetOwns(TypeDBTransaction transaction, AttributeType attributeType) {
        thing_type_unset_owns(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((AttributeTypeImpl) attributeType).concept);
    }

    @Override
    public final void unsetPlays(TypeDBTransaction transaction, RoleType roleType) {
        thing_type_unset_plays(((ConceptManagerImpl) transaction.concepts()).transaction, concept, ((RoleTypeImpl) roleType).concept);
    }

    @Override
    public final String getSyntax(TypeDBTransaction transaction) {
        return thing_type_get_syntax(((ConceptManagerImpl) transaction.concepts()).transaction, concept);
    }

    public static class Root extends ThingTypeImpl {
        public Root(com.vaticle.typedb.client.jni.Concept concept) {
            super(concept);
        }

        @Override
        public final Label getLabel() {
            return Label.of(TypeQLToken.Type.THING.toString());
        }

        @Override
        public boolean isDeleted(TypeDBTransaction transaction) {
            return false;
        }

        @Override
        public ThingTypeImpl getSupertype(TypeDBTransaction transaction) {
            return null;
        }

        @Override
        public Stream<? extends ThingTypeImpl> getSupertypes(TypeDBTransaction transaction) {
            return null;
        }

        @Override
        public Stream<? extends ThingTypeImpl> getSubtypes(TypeDBTransaction transaction) {
            return null;
        }

        @Override
        public Stream<? extends ThingTypeImpl> getSubtypesExplicit(TypeDBTransaction transaction) {
            return null;
        }

        @Override
        public Stream<? extends ThingImpl> getInstances(TypeDBTransaction transaction) {
            return null;
        }

        @Override
        public Stream<? extends ThingImpl> getInstancesExplicit(TypeDBTransaction transaction) {
            return null;
        }
    }
}

File: concept/type/ThingTypeImpl.ts
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Type as TypeProto } from "typedb-protocol/common/concept_pb";
import { Thing } from "../../api/concept/thing/Thing";
import { AttributeType } from "../../api/concept/type/AttributeType";
import { RoleType } from "../../api/concept/type/RoleType";
import { ThingType } from "../../api/concept/type/ThingType";
import { TypeDBTransaction } from "../../api/connection/TypeDBTransaction";
import { ErrorMessage } from "../../common/errors/ErrorMessage";
import { TypeDBClientError } from "../../common/errors/TypeDBClientError";
import { Label } from "../../common/Label";
import { RequestBuilder } from "../../common/rpc/RequestBuilder";
import { Stream } from "../../common/util/Stream";
import { AttributeTypeImpl, EntityTypeImpl, RelationTypeImpl, RoleTypeImpl, ThingImpl, TypeImpl } from "../../dependencies_internal";
import BAD_ENCODING = ErrorMessage.Concept.BAD_ENCODING;

export class ThingTypeImpl extends TypeImpl implements ThingType {

    constructor(name: string, root: boolean) {
        super(Label.of(name), root);
    }

    protected get className(): string {
        return "ThingType";
    }

    asRemote(transaction: TypeDBTransaction): ThingType.Remote {
        return new ThingTypeImpl.Remote(transaction as TypeDBTransaction.Extended, this.label, this.root);
    }

    isThingType(): boolean {
        return true;
    }

    asThingType(): ThingType {
        return this;
    }
}

export namespace ThingTypeImpl {

    export function of(thingTypeProto: TypeProto) {
        if (!thingTypeProto) return null;
        switch (thingTypeProto.getEncoding()) {
            case TypeProto.Encoding.ENTITY_TYPE:
                return EntityTypeImpl.of(thingTypeProto);
            case TypeProto.Encoding.RELATION_TYPE:
                return RelationTypeImpl.of(thingTypeProto);
            case TypeProto.Encoding.ATTRIBUTE_TYPE:
                return AttributeTypeImpl.of(thingTypeProto);
            case TypeProto.Encoding.THING_TYPE:
                return new ThingTypeImpl(thingTypeProto.getLabel(), thingTypeProto.getRoot());
            default:
                throw new TypeDBClientError(BAD_ENCODING.message(thingTypeProto.getEncoding()));
        }
    }

    export class Remote extends TypeImpl.Remote implements ThingType.Remote {

        constructor(transaction: TypeDBTransaction.Extended, label: Label, root: boolean) {
            super(transaction, label, root);
        }

        protected get className(): string {
            return "ThingType";
        }

        asRemote(transaction: TypeDBTransaction): ThingType.Remote {
            return new ThingTypeImpl.Remote(transaction as TypeDBTransaction.Extended, this.label, this.root);
        }

        isThingType(): boolean {
            return true;
        }

        asThingType(): ThingType.Remote {
            return this;
        }

        getSubtypes(): Stream<ThingType> {
            return super.getSubtypes() as Stream<ThingType>;
        }

        getSupertype(): Promise<ThingType> {
            return super.getSupertype() as Promise<ThingType>;
        }

        getSupertypes(): Stream<ThingType> {
            return super.getSupertypes() as Stream<ThingType>;
        }

        getInstances(): Stream<Thing> {
            const request = RequestBuilder.Type.ThingType.getInstancesReq(this.label);
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getThingTypeGetInstancesResPart().getThingsList()))
                .map((thingProto) => ThingImpl.of(thingProto));
        }

        getOwns(): Stream<AttributeType>;
        getOwns(valueType: AttributeType.ValueType): Stream<AttributeType>;
        getOwns(keysOnly: boolean): Stream<AttributeType>;
        getOwns(valueType: AttributeType.ValueType, keysOnly: boolean): Stream<AttributeType>;
        getOwns(valueTypeOrKeysOnly?: AttributeType.ValueType | boolean, keysOnly?: boolean): Stream<AttributeType> {
            let request;
            if (!valueTypeOrKeysOnly) {
                request = RequestBuilder.Type.ThingType.getOwnsReq(this.label, false);
            } else if (typeof valueTypeOrKeysOnly === "boolean") {
                request = RequestBuilder.Type.ThingType.getOwnsReq(this.label, valueTypeOrKeysOnly as boolean)
            } else if (!keysOnly) {
                request = RequestBuilder.Type.ThingType.getOwnsByTypeReq(
                    this.label, (valueTypeOrKeysOnly as AttributeType.ValueType).proto(), false
                );
            } else {
                request = RequestBuilder.Type.ThingType.getOwnsByTypeReq(
                    this.label, (valueTypeOrKeysOnly as AttributeType.ValueType).proto(), keysOnly
                );
            }
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getThingTypeGetOwnsResPart().getAttributeTypesList()))
                .map((attributeTypeProto) => AttributeTypeImpl.of(attributeTypeProto));
        }


        getOwnsExplicit(): Stream<AttributeType>;
        getOwnsExplicit(valueType: AttributeType.ValueType): Stream<AttributeType>;
        getOwnsExplicit(keysOnly: boolean): Stream<AttributeType>;
        getOwnsExplicit(valueType: AttributeType.ValueType, keysOnly: boolean): Stream<AttributeType>;
        getOwnsExplicit(valueTypeOrKeysOnly?: AttributeType.ValueType | boolean, keysOnly?: boolean): Stream<AttributeType> {
            let request;
            if (!valueTypeOrKeysOnly) {
                request = RequestBuilder.Type.ThingType.getOwnsExplicitReq(this.label, false);
            } else if (typeof valueTypeOrKeysOnly === "boolean") {
                request = RequestBuilder.Type.ThingType.getOwnsExplicitReq(this.label, valueTypeOrKeysOnly as boolean)
            } else if (!keysOnly) {
                request = RequestBuilder.Type.ThingType.getOwnsExplicitByTypeReq(
                    this.label, (valueTypeOrKeysOnly as AttributeType.ValueType).proto(), false
                );
            } else {
                request = RequestBuilder.Type.ThingType.getOwnsExplicitByTypeReq(
                    this.label, (valueTypeOrKeysOnly as AttributeType.ValueType).proto(), keysOnly
                );
            }
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getThingTypeGetOwnsExplicitResPart().getAttributeTypesList()))
                .map((attributeTypeProto) => AttributeTypeImpl.of(attributeTypeProto));
        }

        async getOwnsOverridden(attributeType: AttributeType): Promise<AttributeType> {
            const req = RequestBuilder.Type.ThingType.getOwnsOverriddenReq(this.label, ThingType.proto(attributeType));
            return this.execute(req)
                .then((res) => AttributeTypeImpl.of(res.getThingTypeGetOwnsOverriddenRes().getAttributeType()));
        }

        async setOwns(attributeType: AttributeType): Promise<void>;
        async setOwns(attributeType: AttributeType, isKey: boolean): Promise<void>;
        async setOwns(attributeType: AttributeType, overriddenType: AttributeType): Promise<void>;
        async setOwns(attributeType: AttributeType, overriddenTypeOrIsKey?: AttributeType | boolean, isKey?: boolean): Promise<void> {
            let request;
            if (!overriddenTypeOrIsKey) {
                request = RequestBuilder.Type.ThingType.setOwnsReq(this.label, ThingType.proto(attributeType), false);
            } else if (typeof overriddenTypeOrIsKey === "boolean") {
                request = RequestBuilder.Type.ThingType.setOwnsReq(this.label, ThingType.proto(attributeType), overriddenTypeOrIsKey as boolean)
            } else if (!isKey) {
                request = RequestBuilder.Type.ThingType.setOwnsOverriddenReq(
                    this.label, ThingType.proto(attributeType), ThingType.proto((overriddenTypeOrIsKey as AttributeType)), false
                );
            } else {
                request = RequestBuilder.Type.ThingType.setOwnsOverriddenReq(
                    this.label, ThingType.proto(attributeType), ThingType.proto(overriddenTypeOrIsKey as AttributeType), isKey
                );
            }
            await this.execute(request);
        }

        async unsetOwns(attributeType: AttributeType): Promise<void> {
            const request = RequestBuilder.Type.ThingType.unsetOwnsReq(this.label, ThingType.proto(attributeType));
            await this.execute(request);
        }

        getPlays(): Stream<RoleType> {
            const request = RequestBuilder.Type.ThingType.getPlaysReq(this.label);
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getThingTypeGetPlaysResPart().getRoleTypesList()))
                .map((roleProto) => RoleTypeImpl.of(roleProto));
        }

        getPlaysExplicit(): Stream<RoleType> {
            const request = RequestBuilder.Type.ThingType.getPlaysExplicitReq(this.label);
            return this.stream(request)
                .flatMap((resPart) => Stream.array(resPart.getThingTypeGetPlaysExplicitResPart().getRoleTypesList()))
                .map((roleProto) => RoleTypeImpl.of(roleProto));
        }

        async getPlaysOverridden(role: RoleType): Promise<RoleType> {
            const request = RequestBuilder.Type.ThingType.getPlaysOverriddenReq(this.label);
            return this.execute(request)
                .then((res) => RoleTypeImpl.of(res.getThingTypeGetPlaysOverriddenRes().getRoleType()));
        }

        async setPlays(role: RoleType): Promise<void>;
        async setPlays(role: RoleType, overriddenType?: RoleType): Promise<void> {
            let request;
            if (!overriddenType) {
                request = RequestBuilder.Type.ThingType.setPlaysReq(this.label, RoleType.proto(role));
            } else {
                request = RequestBuilder.Type.ThingType.setPlaysOverriddenReq(this.label, RoleType.proto(role), RoleType.proto(overriddenType));
            }
            await this.execute(request);
        }

        async unsetPlays(role: RoleType): Promise<void> {
            const request = RequestBuilder.Type.ThingType.unsetPlaysReq(this.label, RoleType.proto(role));
            await this.execute(request);
        }

        async setAbstract(): Promise<void> {
            const request = RequestBuilder.Type.ThingType.setAbstractReq(this.label);
            await this.execute(request);
        }

        async unsetAbstract(): Promise<void> {
            const request = RequestBuilder.Type.ThingType.unsetAbstractReq(this.label);
            await this.execute(request);
        }

        async isDeleted(): Promise<boolean> {
            return (await this.transaction.concepts.getThingType(this.label.name)) != null;
        }

        protected async setSupertype(thingType: ThingType): Promise<void> {
            const request = RequestBuilder.Type.ThingType.setSupertypeReq(this.label, ThingType.proto(thingType));
            await this.execute(request);
        }
    }
}
